Use Replicas For Read

Cardmarket-Sijing opened the pull request:
This is a proposition on how load could be balanced for replicated clusters.

See:
#595 #392 #173 #579

* Added separate pool and slot map for cluster replicas
* Use replicas to serve read requests

This adds a dependency on the replication strategy for detecting read commands. The detection could be moved to the redis cluster strategy or to a separate class included in both strategies.

Also this doesn't allow for the user to configure how the replicas should be used. Future improvements could be:
* prefer replicas for reads, fallback to primary
* randomly select between replicas and primaries to evenly distribute

This splits up the slots of the primary between replicas. Another idea would be to share the slots between all replicas and select the replica (randomly) for each command

vladvildanov commented on src/Connection/Cluster/RedisCluster.php:
```diff
@@ -75,6 +79,9 @@ public function __construct(
         $this->strategy = $strategy ?: new RedisClusterStrategy();
         $this->slotmap = new SlotMap();
+
+        $this->replicaSlotmap = new SlotMap();
```
Why should we use another SlotMap if we have a SlotMap above?

Cardmarket-Sijing replied:
The current slot map only tracks primaries and does not seem to have functionality for tracking lists of connections per slot. 
I did not want to introduce too many changes at once

vladvildanov commented:
@tillkruss @Cardmarket-Sijing Looks good, but I have few suggestions:

1. Make it configurable. Load balancing should be configurable from client configuration. I suggest to add a support for new option that enables it. `ReplicationStrategy::loadBalancing` should help you:

```
$options = ['cluster' => 'redis, 'loadBalancing' => true];
```

2. Test coverage. We need unit tests and feature tests coverage against cluster.
3. Documentation. Load balancing feature should be documented

vladvildanov commented:
@tillkruss And since this is a breaking change feature we should include it in 3.x

Cardmarket-Sijing commented:
> @tillkruss And since this is a breaking change feature we should include it in 3.x

It could probably be changed to a non breaking change by making it configurable as suggested above and disabled by default.

tillkruss commented:
> It could probably be changed to a non breaking change by making it configurable as suggested above and disabled by default.

Yeah, I'll leave that decision up to you. Configuration would be nice.

Cardmarket-Sijing commented:
I've changed it to be configurable through 

`$options = ['cluster' => 'redis, 'loadBalancing' => true];`

And updated the readme accordingly.

Also added some coverage tests. Let me know if there should be additional test cases

vladvildanov commented on tests/Predis/Connection/Cluster/RedisClusterTest.php:
```diff
@@ -1455,4 +1458,202 @@ public function testQueryClusterNodeForSlotMapPauseDurationOnRetry()
 
         $this->assertCount(16384, $cluster->getSlotMap());
     }
+
```
Mark tests, with `@group disconnected`, represents unit tests

vladvildanov commented:
@Cardmarket-Sijing Looks great! The only missing piece is an integration tests against cluster. Here's an instruction how to perform them locally:

1. Go to directory `docker/unstable_cluster` (make sure you have docker and docker-compose installed) and run `docker-compose up -d`.
2. Write integration tests within the same test class and mark them with annotations: `@group connected`, `@group cluster`
3. Now you can run them and they will be executed against cluster

Cardmarket-Sijing commented:
Something broke while trying to rebase to the main branch, so I made another PR targeting the 2.x branch again.
#1414
